### ISSUE TITLE
Expand window vertically upon double-clicking window border

### ISSUE THREAD

**Issue by Nuc1eoN** [WayfireWM/wayfire#1419]
So this is the one feature that I actually miss from Windows 7/10 and never saw implemented on Linux.

On Windows when the user double-clicks the top or bottom window border (while the resize cursor is shown) the window get maximized vertically (top and bottom get expanded).

I think this feature would be really nice to have, and one that I am still missing on linux after many years.

**Comment by unrelentingtech** [comment]
You could make a plugin that replaces the (client-initiated) maximization with this kind of thing:

```cpp
#define WLR_USE_UNSTABLE
#define WAYFIRE_PLUGIN
#include <wayfire/core.hpp>
#include <wayfire/plugin.hpp>
#include <wayfire/output.hpp>
#include <wayfire/view.hpp>
#include <wayfire/workspace-manager.hpp>
#include <wayfire/signal-definitions.hpp>

class winmax : public wf::plugin_interface_t {
  void init() override {
    output->connect_signal("view-tile-request", &on_maximize_signal);
  };

  wf::signal_connection_t on_maximize_signal = [=](wf::signal_data_t *ddata) {
    auto data = static_cast<wf::view_tile_request_signal *>(ddata);
    auto workspace_impl = output->workspace->get_workspace_implementation();

    if (data->carried_out || !workspace_impl->view_movable(data->view) ||
        !workspace_impl->view_resizable(data->view) ||
        (data->desired_size.width <= 0) || data->edges != wf::TILED_EDGES_ALL) {
      return;
    }

    auto wsg = output->workspace->get_workarea();

    data->desired_size = data->view->get_wm_geometry();
    data->desired_size.y = wsg.y;
    data->desired_size.height = wsg.height;
    data->view->set_geometry(data->desired_size);
    data->carried_out = true; // we do it ourselves instead of just altering edges to let wlr_view_t report a maximized state to xdg-shell
  };
};

DECLARE_WAYFIRE_PLUGIN(winmax);
```

Plus I guess the same thing for the server-side decoration if you care (I don't use SSDs at all).

Now the problem is that this completely replaces the xdg-shell maximized state, if you do care about maximize buttons.. well they'll do this same thing too :D But the `move` snapping / `grid` plugin will still actually maximize to the full screen size.

If you want both this and maximize buttons to work… time to propose an additional stretched-vertically state for xdg-shell in [wayland-protocols](https://gitlab.freedesktop.org/wayland/wayland-protocols) I guess ;)

**Comment by Nuc1eoN** [comment]
@unrelentingtech Oh I see thank you for the broad evaluation of the issue!

Frankly I think, replacing the maximize functionality would miss the point... The intention is to increase user experience not to replace one thing for the other :D
Which is exactly why on Windows this feature is so nice, because it merely enhances already present functionality and builds up on it^^

I see it as a distinction, to simply double click the top bar and to double click while the cursor is in "resize-mode" (targeting the mere border of the window) and double clicking the top bar.

> If you want both this and maximize buttons to work… time to propose an additional stretched-vertically state for xdg-shell in wayland-protocols I guess ;)

So I guess this route is the most viable :D
So do I understand correctly that currently wayland cannot distinguish between double clicking the window border and the top bar?

**Comment by ammen99** [comment]
> So do I understand correctly that currently wayland cannot distinguish between double clicking the window border and the top bar?

For clients with client-side decorations, yes, we cannot.

**Comment by soreau** [comment]
You might be able to guess, by comparing the pointer coordinates with the window geometry? However you would have to make some assumptions about the geometry of the CSD within the client buffer, so it would be a bit of a hack.

**Comment by Nuc1eoN** [comment]
@ammen99 
> For clients with client-side decorations, yes, we cannot.

Let me understand this correctly, so we weed a protocol extension for both CSD and SSD?
Once this extension is added it would then be possible for CSD, but with SSD it is already be possible?

Sorry if anything of this should be obvious, I am totally new to the wayland ecosystem.

**Comment by unrelentingtech** [comment]
> You might be able to guess, by comparing the pointer coordinates with the window geometry?

Yeah, and more importantly tracking the events, e.g. first click in the upper 30% of the surface → second click in the upper 30% of the surface → maximize request for the surface, with nothing significant in between, maybe even with time checks between the events too.

Of course it's a hack but it's not a terrible one. I can't think of anything else being detected by this sequence (maximize buttons are usually single clicked, context menus aren't even the same surface and are also single clicked…)

**Comment by Vladimir-csp** [comment]
In any case, can separate horizontal/vertical maximiziation toggle be added as an action for bindings and window rules?

**Comment by oredaze** [comment]
I would very much like to have vertical maximization. With hotkey however, not clicking tiny borders.
I am good at counter-strike, but that requires effort :)

**Comment by ammen99** [comment]
It is rather easy to achieve this with an IPC script (and you can add this script as a binding with the command plugin).

**Comment by soreau** [comment]
> It is rather easy to achieve this with an IPC script (and you can add this script as a binding with the command plugin).

I suppose you'd want to set the corresponding tiled edges. Is manual resize needed after setting tiled edges?

**Comment by ammen99** [comment]
> Is manual resize needed after setting tiled edges?

Yes

**Comment by ammen99** [comment]
#3040 adds support for vertical and/or horizontal maximization when clicking on the built-in decoration plugin. For other decorator plugins or for client-side decorations, users may try to create custom IPC scripts or write custom plugins, but nothing is planned upstream.
